org.eclipse.xtext/src/org/eclipse/xtext/xtext/FlattenedGrammarAccess.java

```diff
@@ -0,0 +1,332 @@
+
+import org.eclipse.emf.common.notify.Adapter;
+import org.eclipse.emf.common.notify.Notifier;
+import org.eclipse.emf.common.notify.impl.AdapterImpl;
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.EReference;
+import org.eclipse.emf.ecore.util.EcoreUtil;
+import org.eclipse.xtext.AbstractElement;
+import org.eclipse.xtext.AbstractRule;
+import org.eclipse.xtext.CompoundElement;
+import org.eclipse.xtext.Condition;
+import org.eclipse.xtext.EnumRule;
+import org.eclipse.xtext.Grammar;
+import org.eclipse.xtext.GrammarUtil;
+import org.eclipse.xtext.Group;
+import org.eclipse.xtext.Parameter;
+import org.eclipse.xtext.ParserRule;
+import org.eclipse.xtext.RuleCall;
+import org.eclipse.xtext.TerminalRule;
+import org.eclipse.xtext.TypeRef;
+import org.eclipse.xtext.XtextPackage;
+import org.eclipse.xtext.xbase.lib.ListExtensions;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Sets;
+
+/**
+ * @author Sebastian Zarnekow - Initial contribution and API
+ */
+public class FlattenedGrammarAccess {
+	private final Grammar flattenedGrammar;
+
+	public FlattenedGrammarAccess(RuleNames names, RuleFilter filter) {
+		Grammar grammar = names.getContextGrammar();
+		Grammar flattenedGrammar = copy(grammar);
+		flattenedGrammar.setName(grammar.getName());
+		Map<RuleWithParameterValues, AbstractRule> origToCopy = new LinkedHashMap<>();
+		List<AbstractRule> copies = copyRuleStubs(names, origToCopy, filter.getRules(grammar),
+				filter.isDiscardRuleTypeRef());
+		addAll(flattenedGrammar.getRules(), copies);
+		Multimap<TerminalRule, AbstractRule> calledFrom = copyRuleBodies(copies, origToCopy);
+		setHiddenTokens(flattenedGrammar, grammar, origToCopy);
+		markAsFragment(calledFrom);
+		if (filter.isDiscardUnreachableRules()) {
+			Set<AbstractRule> usedRules = new HashSet<>();
+			if (!filter.isDiscardTerminalRules())
+				usedRules.addAll(GrammarUtil.allTerminalRules(flattenedGrammar));
+			UsedRulesFinder finder = new UsedRulesFinder(usedRules);
+			finder.compute(flattenedGrammar);
+			flattenedGrammar.getRules().retainAll(usedRules);
+		}
+		this.flattenedGrammar = flattenedGrammar;
+		new OriginalGrammar(grammar).attachToEmfObject(flattenedGrammar);
+	}
+
+	private void setHiddenTokens(Grammar copy, Grammar orig, Map<RuleWithParameterValues, AbstractRule> origToCopy) {
+		if (orig == null) {
+			copy.setDefinesHiddenTokens(true);
+			return;
+		}
+		if (!orig.isDefinesHiddenTokens()) {
+			setHiddenTokens(copy, head(orig.getUsedGrammars()), origToCopy);
+			return;
+		}
+		copy.setDefinesHiddenTokens(true);
+		addAll(copy.getHiddenTokens(), ListExtensions.map(orig.getHiddenTokens(),
+				hidden -> origToCopy.get(new RuleWithParameterValues(hidden))));
+	}
+
+	private void markAsFragment(Multimap<TerminalRule, AbstractRule> calledFrom) {
+		filter(filter(filter(calledFrom.keySet(), it -> !it.isFragment()),
```

Comment:
Meh... 
FluentIterable.from(..).filter(..).filter(..) would be easier to read.

Reply:
Will do .... please review that I have done it right.